pex/common.py

```diff
@@ -464,10 +464,25 @@ def unlock():
             os.path.join(head, ".{}.atomic_directory.lck".format(tail or "here")),
             os.O_CREAT | os.O_WRONLY,
         )
-        # N.B.: Since lockf operates on an open file descriptor and these are guaranteed to be
-        # closed by the operating system when the owning process exits, this lock is immune to
-        # staleness.
-        fcntl.lockf(lock_fd, fcntl.LOCK_EX)  # A blocking write lock.
+        while True:
+            try:
+                # N.B.: Since lockf operates on an open file descriptor and these are guaranteed to be
+                # closed by the operating system when the owning process exits, this lock is immune to
+                # staleness.
+                fcntl.lockf(lock_fd, fcntl.LOCK_EX)  # A blocking write lock.
+            except OSError as e:
+                deadlock_avoided_errorno = (
+                    errno.EDEADLK  # type: ignore[attr-defined] # See https://github.com/python/typeshed/issues/7551
+                )
+                if e.errno == deadlock_avoided_errorno:
+                    # Another thread/process is doing the work. We must wait.
+                    # N.B.: This can happen if the process running this code forks, this is the same
```

Comment:
Did you mean https://bugs.python.org/issue6721? That appears to have ~nothing to do with this case. It's about mutex style locks and about hanging. We have here, on the other hand, an advisory file lock, and not hanging. To boot, sans a VCS requirement, 0 identified forks in the life-span of the thread pool.

So, something is obviously broken, but I contend we're still clueless what and how. If you agree, I'd prefer a non-misleading comment to the effect, "No clue, but we're deeming moving forward more important than grokking and this seems mostly harmless because ...?". I think I agree mostly harmless involves capping the loop at some maximum time or tries or both. I have no clue how to justify values for those limits though, since I at least, have no clue what's going on here still.

Reply:
The GitHub issue should have enough context, and we can keep it open until you feel it contains enough info regarding this behavior to close it. I linked it instead of the existing NB comment.